### DISCUSSION TITLE
Data Access Request flow

### DISCUSSION THREAD

**Discussion by JensRoland** [open-metadata/OpenMetadata#639] [Ideas]
I would like to propose a data access request/grant flow:

Imagine a data scientist, Sandra, goes to the metadata catalog and finds a table owned by the Finance department. The data in this table is highly sensitive, so the data scientist does not have permission to access it, not even a sample (unless a washed/masked sample is made available). She can tell from the schema that it likely contains the data she needs for her analysis, so she clicks a “Request access”
button right there in the catalog.

Sandra fills in a short request form:

- What is the purpose of the access?
- Who should be granted access? Self, multiple people, a functional group?
- What is the duration of the access? A day, a week, a month?
- Who are the recipients of any derived data / resulting analysis?

This request now appears in her Data Access Requests list, and the data owner is notified of the request. The data owner can now open the Request details page and either:

- Grant (with a duration)
- Reject (w. a reason)
- Ask for details in a “discussion thread” below the request details

Once a request is granted, the metadata platform would need a way to enforce it (or alternatively, a way to allow an external system to do so); I see a few options:

- Connect to the affected stacks and update the vendor specific ACLs / policies to reflect the new grants (primitive, but doable)
- Act as a token vending machine
- Expose an endpoint (SAML 2.0 or another standard?) allowing downstream applications to authorize requests

The reason for implementing the above flow in the metadata catalog should be obvious; no other system contains data about data assets, streaming topics, and even dashboards, as well as data ownership metadata for all of them. This would also create a useful authorization audit trail for checking exactly who had access to what data asset at any given point - and for what reason.

And while there are other solutions for data access control - Kerberos, Active Directory, Lake Formation - these are all tied to specific vendors / technology stacks which makes them unsuitable for an open standard.

**Comment by JensRoland** [comment]
Copied from the Slack channel, @harshach wrote:

(...) most of orgs/teams have a central policy store such as https://ranger.apache.org or their custom one. Wouldn’t it be beneficial if we do the following workflow

1.	First part of workflow would be the same as you proposed. User comes to metadata and asks for request for access
2.	Owner/Team who owns the dataset gets a notification saying User X is interested in reading the data from table along with additional details of who they are and why they need
3.	Team reviews accepts the request in OpenMetadata
4.	This will generate an event which then gets sent to central policy store like apache ranger or one thats implemented in house. We can make this a pluggable implementation such that other policy stores can be integrated as well 
5.	Benefit I see is we can help streamline request/approval process as you noticed and also enable several teams out there who may have existing policy store already implemented and doesn’t have to re-do this work on metadata itself

**Reply by JensRoland** [reply]
My reply: That makes total sense, it’s a pragmatic separation of responsibilities and it makes for a flexible decoupling which should make it work for very different architectures.

**Reply by mithmatt** [reply]
I was doing some research in related space (policy driven data lifecycle and access control), to see if we should introduce a policy entity in OpenMetadata. #720

Reference to my research: https://docs.google.com/document/d/1svyYBpSNBq3XUhALNCOuS-WTYtHNE-NMnghwszDe2G8/edit#
I found AWS and GCP's tag / policy based access control interesting


This discussion has some great insights that broadens my thought process.


I think the challenge here is two-fold:

1. OpenMetadata UI can show information on which users/teams have access to different data entities/tags it has. This should be modeled somehow as metadata and be ingested from the respective services / policy stores.   -- Enables users to discover if they have access to data entities

2. OpenMetadata UI can have users requesting access for themselves (or on behalf of team) to data entities, or even a group of entities identifiable by custom tags (all Sales tables). Take them through approval process, and propagate this information to the respective policy stores / systems.  -- Enables users to collaborate on data entities

**Comment by dfdf** [comment]
This feature would be awesome for our use case too, we are really struggling to find something that can take care of this "request access workflow"

**Comment by bramdes** [comment]
FYI, also looking for something like this "request access button" + workflow

**Comment by alvespat** [comment]
Definietly usefull feature. 
i was looking for central policy management tool based on policy/ABAC. 
It would be awesome if OpenMetadata offer such kind of feature if we would like Open metadata be the one shop stopper  entry point to data product and assets navigation..

**Comment by ThangDao-Qualgo** [comment]
i am looking for any update about this topic, please someone tell me this feature have a plan in furture
